synapse/models/inet.py

```diff
@@ -3306,6 +3464,267 @@ def getModelDefs(self):
+                    )),
+
+                    ('inet:service:permission:type:taxonomy', {}, ()),
+
+                    ('inet:service:permission', {}, (
+
+                        ('name', ('str', {'onespace': True, 'lower': True}), {
+                            'doc': 'The name of the permission.'}),
+
+                        ('type', ('inet:service:permission:type:taxonomy', {}), {
+                            'doc': 'The type taxonomy for the permissions.'}),
+
+                    )),
+
+                    ('inet:service:rule', {}, (
+
+                        ('permission', ('inet:service:permission', {}), {
+                            'doc': 'The permission which is granted.'}),
+
+                        ('denied', ('bool', {}), {
+                            'doc': 'Set to (true) to denote that the rule is an explicit deny.'}),
+
+                        ('object', ('ndef', {'interface': 'inet:service:object'}), {
+                            'doc': 'The object that the permission controls access to.'}),
+
+                        ('grantee', ('ndef', {'forms': ('inet:service:account', 'inet:service:group')}), {
+                            'doc': 'The user or role which is granted the permission.'}),
+                    )),
+
+                    ('inet:service:session', {}, (
+
+                        ('id', ('str', {'strip': True}), {
+                            'doc': 'The service specific session id.'}),
+
+                        ('account', ('inet:service:account', {}), {
+                            'doc': 'The account which authenticated to create the session.'}),
+
+                        ('period', ('ival', {}), {
+                            'doc': 'The period where the session was valid.'}),
+                    )),
+
+                    ('inet:service:login', {}, (
+
+                        ('method', ('inet:service:login:method:taxonomy', {}), {
+                            'doc': 'The type of authentication used for the login. For example "password" or "multifactor.sms".'}),
+
+                        ('session', ('inet:service:session', {}), {
+                            'doc': 'The session generated or refreshed if the login was successful.'}),
+
+                        ('server', ('inet:server', {}), {
+                            'doc': 'The network address of the server which handled the login.'}),
+
+                        ('server:host', ('it:host', {}), {
+                            'doc': 'The server host which authenticated the login.'}),
+
+                        ('client', ('inet:client', {}), {
+                            'doc': 'The network address of the client from the server perspective.'}),
+
+                        ('client:host', ('it:host', {}), {
+                            'doc': 'The client host which initiated the login.'}),
+
+                        # TODO ndef based auth proto details
+                    )),
+
+                    ('inet:service:message', {}, (
```

Comment:
similar to above - not sure props from inet:service:object make sense in inet:service:message (created/removed/creator/owner - what "object" do these refer to in the context of an inet:service:message?)

Note that one option would be to NOT have inet:service:action use inet:service:object - make them independently "inheritable" so forms could use one / the other / both. (If we wanted a precedence order, Synapse would need to support that).

Reply:
A bit confusing but it seemed better to have them all inherit one interface vs being piecemeal'd out. Happy to discuss.